scio-smb/src/test/java/org/apache/beam/sdk/extensions/smb/SortedBucketSourceTest.java

```diff
@@ -692,6 +684,85 @@ public void testUniqueTupleTagIdSecondaryKey() {
     Assert.assertThrows(IllegalArgumentException.class, () -> sb.and(read2));
   }
 
+  @Test
+  public void testMixedInputTypesInSource() throws Exception {
+    final AvroGeneratedUser avroRecord =
+        AvroGeneratedUser.newBuilder()
+            .setName("foo")
+            .setFavoriteColor("avro-color")
+            .setFavoriteNumber(1)
+            .build();
+    final AvroGeneratedUser parquetRecord =
+        AvroGeneratedUser.newBuilder()
+            .setName("foo")
+            .setFavoriteColor("parquet-color")
+            .setFavoriteNumber(2)
+            .build();
+
+    write(
+        lhsPolicy.forDestination(),
+        new AvroBucketMetadata<String, Void, AvroGeneratedUser>(
+            1,
+            1,
+            String.class,
+            "name",
+            null,
+            null,
+            BucketMetadata.HashType.MURMUR3_32,
+            "bucket-",
+            AvroGeneratedUser.class),
+        ImmutableMap.of(BucketShardId.of(0, 0), ImmutableList.of(avroRecord)),
+        AvroFileOperations.of(AvroGeneratedUser.class));
+
+    write(
+        rhsPolicy.forDestination(),
+        new ParquetBucketMetadata<String, Void, AvroGeneratedUser>(
+            1,
+            1,
+            String.class,
+            "name",
+            null,
+            null,
+            BucketMetadata.HashType.MURMUR3_32,
+            "bucket-",
+            AvroGeneratedUser.class),
+        ImmutableMap.of(BucketShardId.of(0, 0), ImmutableList.of(parquetRecord)),
+        ParquetAvroFileOperations.of(AvroGeneratedUser.class));
+
+    final PCollection<KV<String, CoGbkResult>> output =
+        pipeline.apply(
+            Read.from(
+                new SortedBucketPrimaryKeyedSource<>(
+                    String.class,
+                    ImmutableList.of(
+                        new PrimaryKeyedBucketedInput<>(
+                            new TupleTag<AvroGeneratedUser>(),
+                            ImmutableMap.of(
+                                TestUtils.fromFolder(lhsFolder).toString(),
+                                    KV.of(".avro", AvroFileOperations.of(AvroGeneratedUser.class)),
+                                TestUtils.fromFolder(rhsFolder).toString(),
+                                    KV.of(
+                                        ".parquet",
+                                        ParquetAvroFileOperations.of(AvroGeneratedUser.class))),
+                            null)),
+                    TargetParallelism.max(),
+                    "metrics-key")));
+
+    PAssert.thatMap(output)
```

Comment:
So this test fails with an assertion error:

```
java.lang.IllegalArgumentException: All FileOperations Coders must use the same encoding type; found: [org.apache.beam.sdk.io.AvroGeneratedUser, org.apache.avro.generic.GenericRecord]
```

I'd forgotten that `ParquetAvroFileOperations` treats record classes as GenericRecords by default ... hmm..... will have to think on this one :(

Reply:
Maybe we don't need the Coder check at all